text/0000-update-version-handling.md

```diff
@@ -0,0 +1,225 @@
+| `notion pin yarn`         | `notion pin yarn`         |
+| `notion fetch typescript` | `notion fetch typescript` |
+
+Managing a tool at a specific version, at any level of granularity, now uses `@` to separate the tool name from its version.
+
+| Original                     | Proposed                     |
+| ---------------------------- | ---------------------------- |
+| `notion install node latest` | `notion install node@latest` |
+| `notion pin node 10`         | `notion pin node@10`         |
+| `notion fetch node 10.5`     | `notion fetch node@10.5`     |
+
+This also works for shorthands (currently `latest`, conceivably also tags like `next` or `lts` in the future):
+
+| Original                 | Proposed                  | Status      |
+| ------------------------ | ------------------------ | ------------ |
+| `notion pin node latest` | `notion pin node@latest` | current      |
+| `notion pin node lts`    | `notion pin node@lts`    | current      |
+| `notion pin node next`   | `notion pin node@next`   | hypothetical |
+
+## Installing, fetching, or pinning multiple tools
+
+Previously, installing multiple tools required invoking Notion multiple times. Now, it may be invoked with multiple tools at once.
+
+- **Original:**
+
+    ```ts
+    notion install create-react-app latest
+    notion install typescript 3.4
+    ```
+
+- **Updated:**
+
+    ```
+    notion install create-react-app@latest typescript@3.4
+    ```
+
+There are no impacts to features other than fetching, installing, and pinning.
+
+## Implementation changes
+
+The command line implementation will need to be updated, with the install, fetch, and pin commands now receiving a list of tool specs.
+
+The command line parsing will need to be updated to accept variadic arguments of tool-with-optional-versions, using Clap's [`Arg::multiple`](https://docs.rs/clap/2.32.0/clap/struct.Arg.html#method.multiple) argument, with a custom parser. E.g., for `install`, we might have something like this
+
+```rust
+#[derive(StructOpt)]
+pub(crate) struct Install {
+    /// The tool to install, e.g. `node` or `npm` or `yarn`
+    #[structopt(multiple = true), parse(try_from_str = "ToolSpec::try_from_str")]
+    tools: Vec<ToolSpec>,
+}
+```
+
+Here we assume that the `ToolSpec` will add a `try_from_str` or `try_from_os_str` implementation, with a useful error type—here hand-waved as `String`, but it could be as complex as we need as long as the error type has an implementation of `Into<String>`:
+
+```rust
+impl ToolSpec {
+  fn try_from_str(string: &str) -> Result<ToolSpec, String> {
+    // split apart the tool name and its version
+  }
+}
+```
+
+This is a relatively simple implementation. The only complication is the existence of Node namespaces: `@namespace/package`. This means parsing the version is not as simple as splitting on `"@"`. However, it should still be straightforward using the combination of a regular expression and the existing semver tooling.
```

Comment:
I think we should also consult the code for `npm` and `yarn` to see how they do the splitting, so that we don't miss a corner case that we didn't even know existed. I _believe_, however, that the `@` character only represents a namespace if it's the first character in the name, anywhere else it can be treated as the delimiter between the version and the package name, so hopefully the logic shouldn't be too hard.

Reply:
I believe that’s correct and that’s precisely what I was planning. I’ll add a note indicating that intent.